### PR TITLE
perf(dspark): keep ModelOpt LM head in native NVFP4

### PR DESCRIPTION
**Pull request by PatrickPi1312**
## Summary

Keep the current ModelOpt checkpoint's `lm_head` in its native NVFP4 representation and use one FP32-output dp4a path for AR decode, batched verification, and the DSpark proposal head. This avoids NVFP4 -> BF16 -> Q4_K requantization and scores the draft against the target representation it was trained for.

The change also:

- flattens the NVFP4 dequant grid across `y/z` for vocabulary matrices above CUDA's 65,535-row `grid.y` limit;
- records the full stored row count in native payloads so scoring the 65,536-token draft prefix still finds packed weights after all 248,320 scale rows;
- adds GPU regressions for both large-row dequantization and prefix addressing;
- keeps BF16/FP8 checkpoints on the existing Q4_K fallback, with `SPARKINFER_QWEN38_LMHEAD_NVFP4=0` as an A/B escape hatch.

This update rebases the same optimization onto current `main` (`3e15359`), including the intervening full-depth short-generation batching change (`d3eb7db`). The previous auto-eval measured the old, unrebased `be06dfc` against that newer frontier and found only +1.4%. The updated branch now stacks both independent improvements and drops later experimental commits from the fork branch.

## Proof of speedup

- [x] Tested on **RTX 5090** (`sm_120`)

**Decode tok/s** (end-to-end DSpark decode at the active scored context, same final binary and checkpoint):

| | decode tok/s |
|---|--:|
| before (native head disabled) | 123.2106 |
| after (this PR) | 131.4302 |

This is **+6.67%** on `dspark-decode@4k`. AR also moved from 92.2749 to 92.5615 tok/s (+0.31%), and mean acceptance moved from 1.6538 to 1.7534 (+6.02%).

**Prefill pp tok/s** (not targeted):

| | prefill pp tok/s |
|---|--:|
| before prefill (main) | n/a |
| after prefill (this PR) | n/a |

The active scope's protected end-to-end instrument is `runtime/examples/dspark_tau_check.cpp`; `bench/scripts/bench.sh` does not run the target/draft pair.

```text
GPU: NVIDIA GeForce RTX 5090, sm_120, CUDA 12.8
Target: gittensor-model-hub/Qwen3.8-27B-NVFP4-RTX5090
Draft: RadixArk/Qwen3.8-27B-DSpark
Prompt: 4096 tokens, 128 generated tokens
SPARKINFER_DSPARK_SPEC_REPS=3

# before: same final binary with SPARKINFER_QWEN38_LMHEAD_NVFP4=0
SPEC-REPS 3 reps, 0 differ-from-first, 0 not-lossless-vs-AR
DSPARK tau=1.654  steps=78  tokens=128  decode_s=1.039
DSPARK lossless=YES  matched 128/128  (+2 repeats, 0 failed)
AR     92.27 tok/s  (decode 1.387s, ttft 41.637s)
DSPARK 123.21 tok/s  (decode 1.039s)  = 1.335x AR
METRIC AR_TPS 92.2749
METRIC DSPARK_TPS 123.2106
METRIC MEAN_ACCEPT 1.6538
METRIC LOSSLESS 1
METRIC LOSSLESS_RUNS 3

# after: exact final source, native NVFP4 default
SPEC-REPS 3 reps, 0 differ-from-first, 0 not-lossless-vs-AR
DSPARK tau=1.753  steps=73  tokens=128  decode_s=0.974
DSPARK lossless=YES  matched 128/128  (+2 repeats, 0 failed)
AR     92.56 tok/s  (decode 1.383s, ttft 41.571s)
DSPARK 131.43 tok/s  (decode 0.974s)  = 1.420x AR
METRIC AR_TPS 92.5615
METRIC DSPARK_TPS 131.4302
METRIC MEAN_ACCEPT 1.7534
METRIC LOSSLESS 1
METRIC LOSSLESS_RUNS 3
```

## Correctness and validation

The prior official evaluation of this optimization passed every correctness/no-regression guard: `top1=1.0`, `KL=0.0`, exact losslessness across three runs, AR and acceptance floors, and both Qwen3.8/Qwen3.6 long-context guards. The re-evaluation will rerun those gates against current main.

The exact rebased candidate built successfully on the RTX 5090. Relevant GPU tests:

```text
[PASS] DFlash GDN checkpoints and compact commits equal serial state
large-row NVFP4 dequant: OK
NVFP4 vocab-prefix offset: OK logits=[16.000,0.000]
prefill_nvfp4_gpu_test: OK range=[136.000,136.000]
```

Local policy/harness tests on the final branch:

```text
python3 -m unittest discover -s eval -p 'test_*.py'          # 100 passed
python3 -m unittest discover -s bench/scripts -p 'test_*.py' # 42 passed
git diff --check                                             # clean
```

The repository's Windows attested-binary job currently fails on unchanged main at `runtime/examples/qwen_checkpoint.h:55` because MSVC does not define `S_ISDIR`; this optimization does not touch that line. Linux CI and the sensitive-path guard passed on the previous revision.

### CODE REVIEW

**Comment by skyrocket2026** [comment]
## sparkinfer DSpark auto-eval — `eval-dspark:none`

| metric | value |
|---|---|
| **label** | `eval-dspark:none` |
| scored at | **DSpark speculative decode @ ctx=4k** on the ModelOpt NVFP4 checkpoint |
| **PR DSpark tok/s** | **142.21** |
| **main DSpark tok/s** | **140.21** |
| **speedup vs main** | **1.014×** (+1.4%) |
| PR AR tok/s (floor) | 90.52 |
| main AR tok/s (floor) | 90.67 |
| AR vs main (floor) | -0.1% |
| **DSpark vs AR** | **1.571×** — above 1.0 means speculation finally pays |
| mean accept τ | 1.939 (main 1.910, ceiling 7) |
| accuracy gate | ✅ top1=1.000 (bar >=0.9) · KL=0.0000 (bar <=0.1) |
| losslessness gate | ✅ DSpark matches the AR reference token-for-token, verified across **3** independent runs |
| mean accept τ floor | ✅ 1.9394 vs main 1.9104 (bar ≥95%) |
| qwen3.8 (upstream unsloth) guard @16k | ✅ no regression (decode+prefill) |
| qwen3.6 guard @16k | ✅ no regression (decode+prefill) — decode 451.8 tok/s · prefill 26656 pp |
| PPL PR / main | 3.0198 / 3.0198 |
| Polaris receipt | collected, not signed (no key configured) |
| commit | `be06dfcff` |

within significance gate — not a verified dspark-decode@4k improvement

<sub>Scored on the pinned eval box vs same-box `origin/main`: DSpark speculative decode throughput at ctx=4k on the ModelOpt NVFP4 checkpoint, with the AR reference measured in the same process and the same model load. Both a regression in AR decode and any divergence from the AR token sequence are hard REJECTs — a speculative decoder that is fast because it skips verification is not faster, it is wrong. τ is the lever, and the row above reports it against a block_size of 7. This is informational, not a judgment on your PR: a `none` label just means no measurable DSpark decode@4k speedup was verified, which is expected and fine if that isn't what your change is about. Automated — merge behaviour depends on SPARKINFER_DSPARK_AUTOMERGE.</sub>

**Comment by skyrocket2026** [comment]
## Closed: sparkinfer DSpark auto-eval — `eval-dspark:none`

This PR's Qwen3.8-27B prefill@16k speed measured **1.4%** vs main, with no verified prefill@16k improvement (the only scored dimension) — closing automatically. This bot evaluates every eligible PR in the repo against Qwen3.8-27B's decode AND prefill@128 speed specifically, regardless of what the PR is actually about — a close here isn't a judgment on the PR's purpose, just that it didn't move these particular metrics. Reopen (or open a fresh PR) if you have a fix or a different approach.
